File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []

        def dfs(node):
            if not node:
                return

            # 1. left
            dfs(node.left)

            # 2. visit root
            result.append(node.val)

            # 3. right
            dfs(node.right)

        dfs(root)
        return result

File: Data Structures & Algorithms/path-sum/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} targetSum
     * @return {boolean}
     */
    hasPathSum(root, targetSum) {
        function dfs(node, remaining) {
            if (node === null) {
                return false;
            }

            remaining = remaining - node.val

            if (node.left === null && node.right === null) {
               return remaining === 0;
            }

            return (
                dfs(node.left, remaining) ||
                dfs(node.right, remaining)
            );
        }

        return dfs(root, targetSum);
    }
}
